fedbiomed/fedbiomed
# Feature/end to end test machinery

**PR description**

Please find the PR that creates a basic end-to-end testing machinery. It implements a basic python interface for developers to write end-to-end test. 

## Using pytest for performing end to end tests
In the PR several helper python functions are created in order to: 
- Add dataset to nodes
- Start nodes
- Execute python files 
- Execute ipython files

Please see `tests/end2end/e2e_*.py` as an example usage of pytest test files to write. This example show a basic test scenarios where 2-3 nodes are lunched at the beginning of the test and several experiment are created and executed. 

The experiments are  directly created in the test files. However, it is also possible to create ipython or python files outside of the test file and execute it using the method `execute_script.` within the test method written in the test files that is going to be executed via `pytest`.

Please run following command to see an example of end to end execution. `pytest -rfp -s -v tests/end2end/e2e_*.py` 

### Known issues: 
- Component config files are generated in the root `etc` directory. We were obliged to do it that way since fedbiomed components are not 100% compatible to be lunched from another directory than where it is installed. And this forces us to write cleaner functions to clean component data generated during the end to end test. However, cleaning is not really preferable  for researcher since there is only one default researcher for in Fed-BioMed. I preferred to keep it as it is since it may contain some important history from local runs. Nevertheless, the files generated after experiment execution are deleted. 

- Nodes are lunched in a subprocess and their outputs are listen using background thread. This allows developer to lunch nodes and run experiment using single test file. If there is an error in the process it is propagated to the thread that listens for  subprocess outputs but not to the main thread. Therefore, node error does not stops the test execution. However, since experiment will fail due to stopped node the test will fail during `exp.run()` but not before. I will leave possible improvement for later. 

- If experiments are directly created in the test files (not in a separated python script to be executed using `execute_script()`) `clear_experiment_data` should be executed. Execution will clear exerpiment data and most importantly it will stop gRPC server running in a background thread. 

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.


**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`
